### ISSUE TITLE
Exhaustive pattern matching of objects in switch expresssions should not require a wildcard to avoid a CS8509 warning

### ISSUE THREAD

**Issue by olatoft** [dotnet/roslyn#64497]
**Brief description:**

CS8509 gives a warning if a switch expression does not handle all possible values of its input type. For instance, this code

```c#
bool CoastLineIsLongerThan10Km(Country country) =>
    country switch
    {
        LandLockedCountry => false,
        CountryWithCoastLine countryWithCoastLine => countryWithCoastLine.CoastLineLength > 10,
    };

abstract class Country {}
class LandLockedCountry: Country {}

class CountryWithCoastLine: Country
{
    public int CoastLineLength;
}
```

gives the warning "warning CS8509: The switch expression does not handle all possible values of its input type (it is not exhaustive). For example, the pattern '_' is not covered.", even though all possible values of `Country` are handled. To handle the warning, I have to add a wildcard:

```c#
bool CoastLineIsLongerThan10Km(Country country) =>
    country switch
    {
        LandLockedCountry => false,
        CountryWithCoastLine countryWithCoastLine => countryWithCoastLine.CoastLineLength > 10,
        _ => throw new ArgumentOutOfRangeException()
    };
```

**Languages applicable:**

C#. I'm not familiar with VB, so I'm not sure whether it too has the same issue or not.

**Code example that the analyzer should report:**

Would it be possible to not give a CS8509 warning for the first code snippet, but still give a CS8509 warning in the following case, where not all possible values are handled?

```c#
bool CoastLineIsLongerThan10Km(Country country) =>
    country switch
    {
        LandLockedCountry => false,
    };

abstract class Country {}
class LandLockedCountry: Country {}

class CountryWithCoastLine: Country
{
    public int CoastLineLength;
}
```

**Additional information:**

CS8509 is used for switching on both enums and objects. https://github.com/dotnet/roslyn/pull/47693 introduced CS8524, which gives a warning for not handling unnamed enum values, while not handling named enum values now gives a CS8509 warning. I would like to turn on NoWarn for CS8524, and WarningsAsErrors for CS8509, so that I get an error when I forget to handle a named enum value, and no warning when all named enum values are handled. However, because CS8509 is shared between enums and objects, and because exhaustive pattern matching of objects requires a wildcard, turning on WarningsAsErrors for CS8509 would give me an error for not adding a wildcard when switching on objects. If it's not possible to not give a CS8509 warning when switching on objects, would it be possible to use separate diagnostic IDs for switching on enums and switching on objects?

**Documentation requirements:**

When this analyzer is implemented, it must be documented by following the steps at [Documentation for IDE CodeStyle analyzers](https://github.com/dotnet/roslyn/blob/main/docs/contributing/Documentation%20for%20IDE%20CodeStyle%20analyzers.md).

**Comment by olatoft** [comment]
Opened by accident before I finished writing it.

**Comment by CyrusNajmabadi** [comment]
In this case, your switch is not-exhaustive.  As far as the compiler knows, another assembly may add:

`class OtherCountry : Country { }` and as such `country switch { LandLockedCountry => false, }` is non-exhaustive.  Closing out as the compiler is currently working as expected as per the language.  

Note: if you would like a change in language rules, then those should start as discussions on dotnet/csharplang.

**Comment by olatoft** [comment]
Ah, I see. Would it be possible to have exhaustive switch expressions for internal classes?

Would it be possible to use separate diagnostic IDs for switching on enums and switching on classes, as described in "Additional information"?

Thanks for the tip about starting a discussion on dotnet/csharplang. My end goal is to have an "or" type with properties in some or all subtype, which will give a compile error if and only if not all possible subtypes are handled, without requiring a wildcard, similar to discriminated unions in F#. For input to such a discussion, could you tell if the following would be possible in C# from a compiler perspective?

1. Would it be possible to add an access modifier to classes which says that it's possible to use this class and its subclasses from any assembly, but it's only possible to define new subclasses of this class in the same assembly as this class?

2. Would it be possible to add properties to enum values? For instance

```c#
enum Contry
{
    LandLockedCountry,
    CountryWithCoastLine { int CoastLineLength }
}
```

similar to enums in Rust, and conseptually similar to discriminated unions in F#? Or would such a feature be simpler to implement with a whole new data structure instead of adding features to enums?

**Comment by chtenb** [comment]
@olatoft The "closed" sumtype pattern in C# can be done as follows:

```cs
public abstract record Option<T>
{
    private Option() { }
    public sealed record Nothing<T> : Option<T> { }
    public sealed record Some<T>(T Value) : Option<T> { }
}
```

instantiated as

```
var x = new Option<int>.Nothing<int>();
```

Add some static classes and factory methods to make instantiation more convenient. The key part is that Option ha a private constructor, and all the arms are sealed. This means that there is no way to create any more subclasses beside Nothing and Some. @CyrusNajmabadi  It would make sense for the compiler to recognize this pattern and not require the wildcard match for two reasons:

1. It's mildy annoying to have to write all the dead wildcard arms
2. The main idea of a sumtype is exhaustive case distinction. If a case were added, we'd like to receive compiler errors in all places where the new case is not yet handled. If every switch expression contains a wildcard arm however, this does not work.

**Comment by CyrusNajmabadi** [comment]
Not sure why i was pinged :)   Language requests should go to dotnet/csharplang :)

**Comment by chtenb** [comment]
@CyrusNajmabadi I'm not sure what the definition is of a language request, but it sounds like a change to the language specification, i.e. a new syntax or something. The C# sample code I posted above is already valid C#. It's just that the analyzer doesn't recognize that a switch expression that handles all the arms (i.e. Nothing and Some in the example above) is exhaustive, and so an unwarranted warning is issued.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi I'm not sure what the definition is of a language request, but it sounds like a change to the language specification

Yes.  the language specification would have to be changed here.  Currently, from teh language's perspective, this switch is not exhaustive.

> The C# sample code I posted above is already valid C#. It's just that the analyzer doesn't recognize that a switch expression that handles all the arms

Correct, as per the rules of the language.  The compiler needs to implement those rules, which is why you get the issue.  The lang would have to be updated to state that it knows this is exhaustive :)

**Comment by chtenb** [comment]
I understand, thanks for the explanation!

**Comment by chtenb** [comment]
FYI: looks like it was proposed but rejected: https://github.com/dotnet/csharplang/issues/2228

**Comment by CyrusNajmabadi** [comment]
It was not rejected.  The person who filed that issue closed their own issue :)
